### ISSUE TITLE
Dynamic array class factory

### ISSUE THREAD

**Issue by Tito-Kati** [Chemaclass/typed-arrays#22]
I do not know whether to consider it complementary, related or substitute to https://github.com/Chemaclass/typed-arrays/projects/1#card-56451190

**Comment by JesusValera** [comment]
I like your idea, but I'll create a different class (a factory) whose responsibility is to create the `TypedArray` depending the configuration and/or the items.
I'm not 100% agree about to add this responsibility to the constructor of the class directly. 🤔

**Comment by JesusValera** [comment]
I like your idea, maybe I'll put the `ArrayConfig` object as the first parameter, but we can discuss the implementation details later.

Regarding the inferred-type, looks like a bit tricky but sounds a good challenge. 😃

**Comment by Tito-Kati** [comment]
Actually, what I propose is quite easy to implement, the challenge is to find a solution that allows us to solve some of the problems that I mention.

**Comment by Chemaclass** [comment]
I like this idea pretty much, using the first item type as the `typeToEnforce` by default if non `ArrayConfig` is passed:
```php
$dogBreeds = new TypedArray([
    new DogBreed('Golden Retriever'),
    new DogBreed('German Shepherd'),
    new DogBreed('Beagle'),
]); // 100% inferred example
```
The next problem to solve here would be **how should we use/define the argument's type in a function** for this concrete `$dogBreeds` collection? Simply `TypedArray`? 🤔 
```php
function printDogBreeds(TypedArray $dogBreeds): void
{
  foreach($dogBreeds as $dogBreed) {
    // $dogBreed is a DogBreed, but in this function scope, there is no real type checking for that
  }
}
```

**Comment by Tito-Kati** [comment]
Right, it is certainly a problem is one of the reasons why I wanted to see if you could come up with alternatives or improvements.

For example, we could make public the guard method of the type checking (the one called in the constructor) but it would imply reviewing all the types of the elements of the array in each place where we want to use and also assign the responsibility of calling that method to which place where the array is used (losing the power of native typechecking).

That is also why I excused myself from the possibility of using it in places where the power of typechecking did not matter so much, such as rapid prototyping, etc., or even in places where these arrays are generated and immediately consumed, and are not used at all. throughout the entire application, leaving for the latter case our current option to extend the abstract class.

**Comment by JesusValera** [comment]
> I like this idea pretty much, using the first item type as the `typeToEnforce` by default if non `ArrayConfig` is passed:
> 
> ```
> $dogBreeds = new TypedArray([
>     new DogBreed('Golden Retriever'),
>     new DogBreed('German Shepherd'),
>     new DogBreed('Beagle'),
> ]); // 100% inferred example
> ```
> 
> The next problem to solve here would be **how should we use/define the argument's type in a function** for this concrete `$dogBreeds` collection? Simply `TypedArray`? 🤔

Yes, I would go for the less restrictive one, in this case, `array`, if you want a `map` or `list`, you would need to pass the `TypedArrayConfig` as a parameter with the values you wish to specify.


> ```
> function printDogBreeds(TypedArray $dogBreeds): void
> {
>   foreach($dogBreeds as $dogBreed) {
>     // $dogBreed is a DogBreed, but in this function scope, there is no real type checking for that
>   }
> }
> ```
 
That's true, maybe we would need to return on the fly a new instance of the abstract class like?

```php
return new([$dogs]) extends TypedArray { ... }
```

**Comment by Tito-Kati** [comment]
@JesusValera I had thought to obtain the class of the first element of the array, a reflection is not necessary in any case. Although what you say is totally the idea I had for the implementation. But I think this is not the problem that @Chemaclass mentions.

I understand that @Chemaclass refers to the typecheck of the `printDogBreeds` method, which is accepting **any** `TypedArray` and is not guaranteed to be an array of `DogBreed`s.

**Comment by JesusValera** [comment]
Yes @Tito-Kati, I get your point.
As you said, I see that as prototyping.
If you want something fast, you can use this factory (or whatever), but if you want something more robust with type-checking and so on, you need to create your own typed-array class.
I don't see a problem on this 😜
